clash-lib/src/Clash/Rewrite/Util.hs

```diff
@@ -185,24 +204,36 @@ applyDebug
   -- ^ New expression
   -> RewriteMonad extra Term
 applyDebug name exprOld hasChanged exprNew = do
-  nTrans <- Lens.use transformCounter
+  countersV <- Lens.use transformCounters
+  counters <- MVar.takeMVar countersV
   opts <- Lens.view debugOpts
 
+  let nTrans = sum counters
   let from = fromMaybe 0 (dbg_transformationsFrom opts)
   let limit = fromMaybe maxBound (dbg_transformationsLimit opts)
 
-  if | nTrans - from > limit ->
+  if | nTrans - from > limit -> do
+         MVar.putMVar countersV counters
          error "-fclash-debug-transformations-limit exceeded"
-     | nTrans <= from ->
+     | nTrans <= from -> do
+         MVar.putMVar countersV counters
          pure exprNew
      | otherwise ->
-         go opts
+         go counters (pred nTrans) opts
  where
-  go opts = traceIf (hasDebugInfo TryTerm name opts) ("Tried: " ++ name ++ " on:\n" ++ before) $ do
-    nTrans <- pred <$> Lens.use transformCounter
+  go counters nTrans opts = do
+    ioLockV <- Lens.use ioLock
+
+    MVar.withMVar ioLockV $ \() ->
+      traceWhen (hasDebugInfo TryTerm name opts) ("Tried: " ++ name ++ " on:\n" ++ before)
+
+    countersV <- Lens.use transformCounters
 
-    Monad.when (dbg_countTransformations opts && hasChanged) $ do
-      transformCounters %= HashMap.insertWith (const succ) (Text.pack name) 1
+    Monad.when (dbg_countTransformations opts && hasChanged) $
+      MVar.putMVar countersV (HashMap.insertWith (const succ) (Text.pack name) 1 counters)
```

Comment:
@vmchale Now it's only putting the counters MVar back when there is no change: https://github.com/clash-lang/clash-compiler/blob/6866bcfccbc2bf241394b8bf3be7542cd1782715/clash-lib/src/Clash/Rewrite/Util.hs#L230-L235

Which will create a deadlock when there is a change, because the MVar isn't fill anymore.
Like I said above, I think it's better to not empty the MVar in the first place in `applyDebug`

(Try running the compiler with `-fclash-debug-history`, to see this deadlock)

Reply:
Ok, I think I got something sensible!